Assets/MixedRealityToolkit/Providers/Hands/HandRay/AltHandRay.cs

```diff
@@ -0,0 +1,164 @@
+        private Vector3 palmNormal;
+        private Vector3 headForward;
+        private Vector3 headUp;
+
+        private Vector3 lastStableHeadPose = Vector3.zero;
+        private Vector3 lastTorsoDirection = Vector3.zero;
+
+        // HeadPose stabilization
+        // Threshold distance used to update the tracked head position (10cm - 0.1m).
+        protected const float cHeadDistanceThreshold = 0.1f;
+
+        protected const float MinHeadPoseDistanceForCorrection = cHeadDistanceThreshold / 2f;
+
+        /// <summary>
+        /// Update data used by Hand Implementation, to compute next HandRay.
+        /// </summary>
+        /// <param name="handPosition">Position of hand</param>
+        /// <param name="palmNormal">Palm normal</param>
+        /// <param name="headTransform">Transform of CameraCache.main</param>
+        /// <param name="sourceHandedness">Handedness of related hand</param>
+        public void Update(Vector3 handPosition, Vector3 palmNormal, Transform headTransform, Handedness sourceHandedness)
+        {
+            Vector3 newStableHeadPose = UpdateStableHeadPose(headTransform);
+            MixedRealityPose stabilizedHeadPose = new MixedRealityPose(newStableHeadPose, headTransform.rotation);
+
+            // Approximate new torso direction.
+            Vector3 newTorsoDirection = ComputeApproximateTorsoDirection(stabilizedHeadPose, handPosition, sourceHandedness);
+
+            // Mix the new torso direction with the last to ease into direction changes.
+            lastTorsoDirection = Vector3.Lerp(newTorsoDirection, lastTorsoDirection, 0.5f);
+
+            Quaternion torsoRotation = Quaternion.LookRotation(lastTorsoDirection, Vector3.up);
+
+            Vector3 rayPivotPoint = ComputeApproximateShoulderPosition(stabilizedHeadPose, torsoRotation, sourceHandedness);
+
+            Vector3 handRayDirection = (handPosition - rayPivotPoint).normalized;
+
+            this.palmNormal = palmNormal;
+            this.headForward = headTransform.forward;
+            this.headUp = headTransform.up;
+
+            stabilizedRay.AddSample(new Ray(handPosition, handRayDirection));
+        }
+
+        private Vector3 UpdateStableHeadPose(Transform headTransform)
+        {
+            // We check the distance from the last received head pose, and keep the previous value if the distance hasn't sufficiently changed.
+            // This is done to prevent shifts to the head position caused by slight head rotation.
+            float distanceFromLastPose = Vector3.Distance(headTransform.position, lastStableHeadPose);
+
+            if (distanceFromLastPose > MinHeadPoseDistanceForCorrection)
+            {
+                float headDistanceOffset = Mathf.Clamp01(distanceFromLastPose / cHeadDistanceThreshold);
+                float correctionAmt = Mathf.Pow(headDistanceOffset, 4f);
+                lastStableHeadPose = Vector3.Lerp(lastStableHeadPose, headTransform.position, correctionAmt);
```

Comment:
other minor suggestion, calculate the sqrMagnitude instead of Distance for the if statement comparison, then perform the sqrt() operation in the if statement block if necessary. So that way the expensive sqrt() calculation is only performed when needed

Reply:
This is a good optimization. I'm happy you suggested it. I'll try using it more often in my personal code.